File: 703.数据流中的第K大元素/KthLargest.go
package problem0703

import "sort"

// https://leetcode-cn.com/problems/kth-largest-element-in-a-stream/
// 显然不能保存数据流中的所有数据，而是只保存k个最大的数据，构建小顶堆
type KthLargest struct {
	nums []int
	k    int
}

func Constructor(k int, nums []int) KthLargest {
	if len(nums) > k {
		sort.Ints(nums)
		nums = nums[len(nums)-k:]
	}
	heapify(nums)
	return KthLargest{
		nums: nums,
		k:    k,
	}
}

func (this *KthLargest) Add(val int) int {
	if len(this.nums) < this.k {
		this.nums = append(this.nums, val)
		siftUp(this.nums, len(this.nums)-1)
	} else {
		if val > this.nums[0] {
			this.nums[0] = val
			siftDown(this.nums, 0)
		}
	}
	return this.nums[0]
}

func heapify(nums []int) {
	for last := (len(nums) - 2) / 2; last >= 0; last-- {
		siftDown(nums, last)
	}
}

// 自顶向下调整，用于调整堆顶
func siftDown(nums []int, index int) {
	leftChildIndex := index*2 + 1
	if leftChildIndex >= len(nums) {
		return
	}
	if leftChildIndex+1 < len(nums) && less(nums[leftChildIndex+1], nums[leftChildIndex]) {
		leftChildIndex++
	}
	if less(nums[leftChildIndex], nums[index]) {
		swap(nums, index, leftChildIndex)
		siftDown(nums, leftChildIndex)
	}
}

// 自底向上调整，用于调整整个堆
func siftUp(nums []int, index int) {
	parentIndex := (index - 1) / 2
	if less(nums[index], nums[parentIndex]) {
		swap(nums, index, parentIndex)
		siftUp(nums, parentIndex)
	}
}

func less(a, b int) bool {
	return a < b
}

func swap(nums []int, i, j int) {
	nums[i], nums[j] = nums[j], nums[i]
}

File: 703.数据流中的第K大元素/KthLargest.js
class KthLargest {
  constructor(k, nums) {
    this.k = k
    // 排序后截取最大的k个
    nums.sort((a, b) => a > b ? -1 : 1)
    nums = nums.slice(0, k)
    this.heap = new Heap(nums, (a, b) => a < b)
  }
  // 维护一个大小为k的小顶堆
  add(val) {
    if (this.heap.length < this.k) {
      this.heap.push(val)
    } else {
      if (val > this.heap.peak()) {
        this.heap.push(val)
        this.heap.pop()
      }
    }
    return this.heap.peak()
  }
}


// 堆
class Heap {
  constructor(data = [], less) {
    this.data = data
    this.less = less || ((a, b) => a < b)

    if (this.length) {
      for (let p = (this.length - 2) >> 1; p >= 0; p--) {
        this._down(p)
      }
    }
  }
  get length() {
    return this.data.length
  }
  peak() {
    return this.data[0]
  }
  push(val) {
    this.data.push(val)
    this._up(this.length - 1)
  }
  pop() {
    if (!this.length) {
      return
    }
    this._swap(0, this.length - 1)
    const popItem = this.data.pop()
    this._down(0)
    return popItem
  }
  _swap(i, j) {
    [this.data[i], this.data[j]] = [this.data[j], this.data[i]]
  }
  _up(i) {
    if (i <= 0) return
    const pIndex = (i - 1) >> 1
    if (this.less(this.data[i], this.data[pIndex])) {
      this._swap(i, pIndex)
      this._up(pIndex)
    }
  }
  _down(i) {
    let leftIndex = i * 2 + 1
    if (leftIndex >= this.length) {
      return
    }
    if (leftIndex + 1 < this.length && this.less(this.data[leftIndex+1], this.data[leftIndex])) {
      leftIndex++
    }
    if (this.less(this.data[leftIndex], this.data[i])) {
      this._swap(leftIndex, i)
      this._down(leftIndex)
    }
  }
}
